kinesis/src/main/scala/akka/stream/alpakka/kinesis/impl/ShardProcessor.scala

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2016-2019 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.kinesis.impl
+
+import java.util.concurrent.Semaphore
+
+import akka.annotation.InternalApi
+import akka.stream.alpakka.kinesis.CommittableRecord
+import akka.stream.alpakka.kinesis.CommittableRecord.{BatchData, ShardProcessorData}
+import software.amazon.kinesis.lifecycle.ShutdownReason
+import software.amazon.kinesis.lifecycle.events._
+import software.amazon.kinesis.processor.{RecordProcessorCheckpointer, ShardRecordProcessor}
+import software.amazon.kinesis.retrieval.KinesisClientRecord
+
+import scala.collection.JavaConverters._
+
+@InternalApi
+private[kinesis] class ShardProcessor(
+    callback: CommittableRecord => Unit
+) extends ShardRecordProcessor {
+
+  // We keep track of the last Record of the Shard using a Semaphore instance
+  // The last Record of the Shard must be checkpointed before the shard is "closed"
+  // otherwise some Records might be acknowledged but actually not processed by the stream
+  //
+  // This ShardProcessor instance and its Semaphore will live in a thread spawned by the
+  // AWS KCL Scheduler after the Scheduler run() method has been invoked.
+  private val lastRecordSemaphore = new Semaphore(1)
+
+  private var shardData: ShardProcessorData = _
+  private var checkpointer: RecordProcessorCheckpointer = _
+  private var shutdown: Option[ShutdownReason] = None
+
+  override def initialize(initializationInput: InitializationInput): Unit =
+    shardData = new ShardProcessorData(initializationInput.shardId,
+                                       initializationInput.extendedSequenceNumber,
+                                       initializationInput.pendingCheckpointSequenceNumber)
+
+  override def processRecords(processRecordsInput: ProcessRecordsInput): Unit = {
+    checkpointer = processRecordsInput.checkpointer()
+
+    val batchData = new BatchData(processRecordsInput.cacheEntryTime,
+                                  processRecordsInput.cacheExitTime,
+                                  processRecordsInput.isAtShardEnd,
+                                  processRecordsInput.millisBehindLatest)
+
+    if (batchData.isAtShardEnd) {
+      lastRecordSemaphore.acquire()
+    }
+
+    val numberOfRecords = processRecordsInput.records().size()
+    processRecordsInput.records().asScala.zipWithIndex.foreach {
+      case (record, index) =>
+        callback(
+          new InternalCommittableRecord(
+            record,
+            batchData,
+            lastRecord = processRecordsInput.isAtShardEnd && index + 1 == numberOfRecords
+          )
+        )
+    }
+  }
+
+  final class InternalCommittableRecord(record: KinesisClientRecord, batchData: BatchData, lastRecord: Boolean)
+      extends CommittableRecord(record, batchData, shardData) {
+    private def checkpoint(): Unit = {
+      checkpointer.checkpoint(record.sequenceNumber(), record.subSequenceNumber())
```

Comment:
Is there any risk the `checkpointer` could be `null` here?

Reply:
No, the `checkpointer` is gonna be referenced as soon as the `ShardProcessor` starts processing records (`processRecords` method is invoked). `checkpointer` reference can change but the variable is never de-referenced.